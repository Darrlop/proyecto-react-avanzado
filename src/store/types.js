export const AUTH_LOGIN_PENDING = 'auth/login/pending';
export const AUTH_LOGIN_FULFILLED = 'auth/login/fulfilled';
export const AUTH_LOGIN_REJECTED = 'auth/login/rejected';

export const AUTH_LOGOUT_FULFILLED = 'auth/logout';

export const ADVERTS_LOADED_PENDING = 'adverts/pending';
export const ADVERTS_LOADED_FULFILLED = 'adverts/fulfilled';
export const ADVERTS_LOADED_REJECTED = 'adverts/rejected';

export const ADVERT_DETAIL_PENDING = 'adverts/:advertId/pending';
export const ADVERT_DETAIL_FULFILLED = 'adverts/:advertId/fulfilled';
export const ADVERT_DETAIL_REJECTED = 'adverts/:advertId/rejected';

export const ADVERT_NEW_PENDING = 'adverts/new/pending';
export const ADVERT_NEW_FULFILLED = 'adverts/new/fulfilled';
export const ADVERT_NEW_REJECTED = 'adverts/new/rejected';

export const ADVERT_DELETED_PENDING = 'adverts/:advertId/pending';
export const ADVERT_DELETED_FULFILLED = 'adverts/:advertId/fulfilled';
export const ADVERT_DELETED_REJECTED = 'adverts/:advertId/rejected';

export const TAGS_LOADED_PENDING = 'adverts/tags/pendind';
export const TAGS_LOADED_FULFILLED = 'adverts/tags/fulfilled';
export const TAGS_LOADED_REJECTED = 'adverts/tags/rejected';

export const UI_RESET_ERROR = 'ui/reset-error';