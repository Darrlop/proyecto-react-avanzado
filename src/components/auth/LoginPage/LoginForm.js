import T from 'prop-types';

import useForm from '../../../hooks/useForm';

const validEmail = ({ email }) => email;
const validPassword = ({ password }) => password;

function LoginForm({ onSubmit, isLoading }) {
  const {
    formValue: credentials,
    handleChange,
    handleSubmit,
    validate,
  } = useForm({
    email: '',
    password: '',
    remember: false,
  });
  const { email, password, remember } = credentials;
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <label htmlFor="email">Email:
        <input id="email" name="email" value={email} onChange={handleChange} />
      </label>
      <label htmlFor="password">Password:
        <input
          type="password"
          id="password"
          name="password"
          value={password}
          onChange={handleChange}
        />
      </label>
      <input
        type="checkbox"
        name="remember"
        checked={remember}
        onChange={handleChange}
      />
      <button disabled={!validate(validEmail, validPassword, () => !isLoading)}>
        Login
      </button>
    </form>
  );
}

LoginForm.propTypes = {
  isLoading: T.bool,
  onSubmit: T.func.isRequired,
};

export default LoginForm;
