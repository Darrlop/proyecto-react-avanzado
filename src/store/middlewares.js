export const failureRedirects = (router, redirectsMap) => _store => next => action => {
  const result = next(action);

  // Redundante. Hago return result al final de la funci√≥n
  // if (!action.error){
  //   return result;
  // }

  if (action.error) {
    const redirect = redirectsMap[action.payload.statusCode];
    if (redirect) {
      router.navigate(redirect);
    }
  }

  return result;
};