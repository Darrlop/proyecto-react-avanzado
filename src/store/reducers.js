import { combineReducers } from 'redux';
import {
  AUTH_LOGIN_PENDING,
  AUTH_LOGIN_FULFILLED,
  AUTH_LOGIN_REJECTED,
  AUTH_LOGOUT,
  // ADVERTS_LOADED_PENDING,
  // ADVERTS_LOADED_FULFILLED,
  // ADVERTS_LOADED_REJECTED,
  // ADVERT_DETAIL_PENDING,
  // ADVERT_DETAIL_FULFILLED,
  // ADVERT_DETAIL_REJECTED,
  // ADVERT_NEW_PENDING,
  // ADVERT_NEW_FULFILLED,
  // ADVERT_NEW_REJECTED,
  UI_RESET_ERROR
} from './types'
import { func } from 'prop-types';


//BALIZA falta adverts y ui
export const defaultState = {
  auth: false,
  ui: {
    pending: false,
    error: null,
  }
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_FULFILLED:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
};

//BALIZA falta a√±adir los advets y  luego refactorizar con combineReducers
// export default function reducer(state = defaultState, action) {
//   return {
//     aut: auth(state.auth, action)
//   }
// }


export function ui(state = defaultState.ui, action) {
  switch (action.type) {
    case UI_RESET_ERROR:
      return { ...state, error: null };
    case AUTH_LOGIN_PENDING:
      return { ...state, pending: true };
    case AUTH_LOGIN_FULFILLED:
      return { ...state, pending: false, error: null };
    case AUTH_LOGIN_REJECTED:
      return { ...state, pending: false, error: action.payload };
    default:
      return state;
  }
}