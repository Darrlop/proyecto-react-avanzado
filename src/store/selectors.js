export const getIslogged = state => state.auth;
export const getAreAdvertsLoaded = state => state.adverts.loaded;
//export const getAdverts = state => state.adverts.data;
export const getStateAdverts = state => state.adverts.data;
export const getAdvert = advertId => state => getStateAdverts(state).find(advert => advert.id === advertId);

export const getAreTagsLoaded = state => getTags(state).length > 0;
export const getTags = state => state.tags;

export const getUi = state => state.ui;