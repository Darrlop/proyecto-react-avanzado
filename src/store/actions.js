import {
  AUTH_LOGIN_PENDING,
  AUTH_LOGIN_FULFILLED,
  AUTH_LOGIN_REJECTED,
  AUTH_LOGOUT,
  ADVERTS_LOADED_PENDING,
  ADVERTS_LOADED_FULFILLED,
  ADVERTS_LOADED_REJECTED,
  ADVERT_DETAIL_PENDING,
  ADVERT_DETAIL_FULFILLED,
  ADVERT_DETAIL_REJECTED,
  ADVERT_NEW_PENDING,
  ADVERT_NEW_FULFILLED,
  ADVERT_NEW_REJECTED,
  ADVERT_DELETED_PENDING,
  ADVERT_DELETED_FULFILLED,
  ADVERT_DELETED_REJECTED,
  TAGS_LOADED_PENDING,
  TAGS_LOADED_FULFILLED,
  TAGS_LOADED_REJECTED,
  UI_RESET_ERROR
} from './types'

import { getAdvert, getAreAdvertsLoaded, getAreTagsLoaded } from './selectors';
import { createAdvert } from '../components/adverts/service';


export const authLoginPending = () => ({
  type: AUTH_LOGIN_PENDING,
});

export const authLoginFulfilled = () => ({
  type: AUTH_LOGIN_FULFILLED,
});

export const authLoginRejected = error => ({
  type: AUTH_LOGIN_REJECTED,
  payload: error,
  error: true,
});


//Defino el thunk para login
export const authLogin = credentials => {
  return async function (dispatch, _getState, { services: { auth }, router }) {
    try {
      dispatch(authLoginPending());
      await auth.login(credentials);
      dispatch(authLoginFulfilled());
      const to = router.state.location.state?.from || '/';
      router.navigate(to, { replace: true });
    } catch (error) {
      dispatch(authLoginRejected(error));
      throw error;
    }
  };
};


export const authLogout = () => ({
  type: AUTH_LOGOUT,
});


export const advertsLoadedPending = () => ({
  type: ADVERTS_LOADED_PENDING,
});
export const advertsLoadedFulfilled = adverts => ({
  type: ADVERTS_LOADED_FULFILLED,
  payload: adverts
});
export const advertsLoadedRejected = error => ({
  type: ADVERTS_LOADED_REJECTED,
  payload: error,
  error: true,
});

//Defino el thunk para anuncios
export const loadAdverts = () => {
  return async function (dispatch, getState, { services: { adverts } }) {

    //Detecto si hay anuncios ya cargados para no repetir operaciÃ³n
    const state = getState();
    if (getAreAdvertsLoaded(state)) {
      return;
    }

    try {
      dispatch(advertsLoadedPending());
      const loadedAdverts = await adverts.getAdverts();
      dispatch(advertsLoadedFulfilled(loadedAdverts));
    } catch (error) {
      dispatch(advertsLoadedRejected(error));
      throw error;
    }
  };
};

export const advertDetailPending = () => ({
  type: ADVERT_DETAIL_PENDING,
});
export const advertDetailFulfilled = advert => ({
  type: ADVERT_DETAIL_FULFILLED,
  payload: advert
});
export const advertDetailRejected = error => ({
  type: ADVERT_DETAIL_REJECTED,
  payload: error,
  error: true,
});

//Defino el thunk para una anuncio concreto
export const loadAdvert = advertId => {
  return async function (dispatch, getState, { services: { adverts } }) {
    if (getAdvert(advertId)(getState())) {
      return;
    }
    try {
      dispatch(advertsLoadedPending());
      const loadedAdvertOne = await adverts.getAdverts(advertId);
      dispatch(advertDetailFulfilled(loadedAdvertOne));
    } catch (error) {
      dispatch(advertDetailRejected(error));
      throw error;
    }
  };
};


export const advertsDeletedPending = () => ({
  type: ADVERT_DELETED_PENDING,
});
export const advertsDeletedFulfilled = advert => ({
  type: ADVERT_DELETED_FULFILLED,
  payload: advert
});
export const advertsDeletedRejected = error => ({
  type: ADVERT_DELETED_REJECTED,
  error: true,
  payload: error
});

export const deleteAdvert = advertId => {
  return async function (dispatch, _getState, { services: { adverts }, router }) {
    try {
      dispatch(advertsDeletedPending());
      await adverts.deleteAdvert(advertId);
      dispatch(advertsDeletedFulfilled(advertId));
      router.navigate(`/adverts`);
    } catch (error) {
      dispatch(advertsDeletedRejected(error));
    }
  };
};


export const advertNewPending = () => ({
  type: ADVERT_NEW_PENDING,
});
export const advertNewFulfilled = advert => ({
  type: ADVERT_NEW_FULFILLED,
  payload: advert,
});
export const advertNewRejected = error => ({
  type: ADVERT_NEW_REJECTED,
  error: true,
  payload: error,
});

export const createNewAdvert = newAdvert => {
  return async function (dispatch, _getState, { services: { adverts }, router }) {
    try {
      dispatch(advertNewPending());
      const advert = await adverts.createAdvert(newAdvert);
      dispatch(advertNewFulfilled(advert));
      //router.navigate(`/adverts/${advert.id}`);
      router.navigate(`/adverts`);
    } catch (error) {
      dispatch(advertNewRejected(error));
    }
  };
};


export const tagsLoadedPending = () => ({
  TAGS_LOADED_PENDING
});
export const tagsLoadedFulfilled = tags => ({
  type: TAGS_LOADED_FULFILLED,
  payload: tags,
});
export const tagsLoadedRejected = error => ({
  type: TAGS_LOADED_REJECTED,
  payload: error,
  error: true,
});

export const loadTags = () => {
  return async function (dispatch, getState, { services: { adverts } }) {
    if (getAreTagsLoaded(getState())) {
      return;
    }
    const tags = await adverts.getTags();
    dispatch(tagsLoadedFulfilled(tags));
  };
};


export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
