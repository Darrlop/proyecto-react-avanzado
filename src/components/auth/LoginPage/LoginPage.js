import React from 'react';
import { useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import LoginForm from './LoginForm';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../../store/actions';
import { getUi } from '../../../store/selectors';

function LoginPage() {
  const navigate = useNavigate();
  const location = useLocation();
  const dispatch = useDispatch();

  const resetError = () => dispatch(uiResetError());

  const { pending, error } = useSelector(getUi)
  const isLoading = pending;

  const [formValues, setFormValues] = useState({
    username: '',
    password: '',
  });


  const handleSubmit = credentials => {
    dispatch(authLogin(credentials));
  };

  return (
    <div>
      <LoginForm onSubmit={handleSubmit} isLoading={pending} />
      {isLoading && <p>...loge√°ndose en nodepop</p>}
      {error && (
        <div onClick={resetError} style={{ color: 'red' }}>
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
