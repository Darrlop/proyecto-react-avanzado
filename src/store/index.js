import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import { withExtraArgument } from "redux-thunk";

import * as reducers from './reducers';
import * as actionCreators from './actions';
import * as auth from '../components/auth/service';
import * as adverts from '../components/adverts/service'

//import { failureRedirects } from './middleware';


const reducer = combineReducers(reducers);

// composeEnhancers para habilitar devtools
const composeEnhancers = composeWithDevTools({ actionCreators });


export default function configureStore(preloadedState, { router }) {
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(
      applyMiddleware(
        withExtraArgument({ services: { auth, adverts }, router }),
        // failureRedirects(router, {
        //   401: '/login',
        //   404: '/404',
        // }),
      ),
    ),
  );


  return store;
}