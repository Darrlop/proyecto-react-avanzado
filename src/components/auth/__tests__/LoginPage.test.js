import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LoginPage from "../LoginPage";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { authLogin } from "../../../store/actions";
import { toBeDisabled, toBeEnabled } from "@testing-library/jest-dom";

jest.mock('../../../store/actions');
//Para evitar warnings con esLint
const userType = (input, text) => userEvent.type(input, text);

describe('LoginPage', () => {

  const state = { auth: false, ui: { pending: false, error: null } };
  const store = {
    dispatch: () => { },
    getState: () => state,
    subscribe: () => { },
  };
  const renderComponent = () =>
    render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider >,
    );


  test('Snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();

  });


  test('shouls dispatch authLogin action', () => {

    const email = 'email';
    const password = 'password';
    const remember = false;

    renderComponent();

    const emailInput = screen.getByLabelText(/email/i); // Uso expresiones regulares. La 'i' indica que no es sensible a Mayus/Minus
    const passwordInput = screen.getByLabelText(/password/i);
    const submitButton = screen.getByRole(/button/);

    expect(submitButton).toBeDisabled();
    act(() => userType(emailInput, email));
    act(() => userType(passwordInput, password));

    expect(submitButton).toBeEnabled();
    userEvent.click(submitButton);

    // expect(authLogin).toHaveBeenCalledWith({ checkSession, username, password });
    expect(authLogin).toHaveBeenCalledWith({ email, password, remember });
  });

});