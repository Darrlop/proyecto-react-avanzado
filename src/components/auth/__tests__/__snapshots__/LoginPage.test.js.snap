// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Snapshot 1`] = `
<div>
  <div>
    <form>
      <label
        for="email"
      >
        Email:
        <input
          id="email"
          name="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password:
        <input
          id="password"
          name="password"
          type="password"
          value=""
        />
      </label>
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
